### ISSUE TITLE
 CatalogApi entry in AuthProviderFactoryOptions adds unnecessary coupling 

### ISSUE THREAD

**Issue by GoWind** [backstage/backstage#3867]
I am not sure how I missed this change, but I  would have provided feedback when this was done.

The AuthProviderFactory options now has a `CatalogApi` [field](https://github.com/backstage/backstage/blob/master/plugins/auth-backend/src/providers/types.ts#L122), whose value is any instance implementing the `CatalogApi` interface. 

The [CatalogApi](https://github.com/backstage/backstage/blob/master/packages/catalog-client/src/CatalogClient.ts) is a rather complex interface with multiple non-optional methods that are to be implemented. 
This makes it more complicated for anyone trying to implement their own AuthProviderFactory. To me, the motivation for adding this field to AuthProviderFactory is rather unclear and or partial (based on the [issue](https://github.com/backstage/backstage/pull/3298)).

I assume that you want to encapsulate a client to the catalog backend in such a way that requests are authenticated? , or that users are validated using the catalog. 

As for the former, I think, this is the wrong place to add authentication to the catalog. The Authentication Provider should not be coupled tightly to the catalog client , but should rather allow some sort of mechanism to verify an incoming request before passing on the request to the catalog's request handler. 

As to the latter, the only place I see the catalog used to populate the Identity of the user during the authentication process is when a user uses Google's Authentication [here](https://github.com/backstage/backstage/blob/master/plugins/auth-backend/src/providers/google/provider.ts#L21)

This assumption is incorrect, as it assumes that if a user's profile is not part of the response  , then the user is present in the Catalog. This assumption by itself isn't a problem, but the consequence of this assumption is that **it couples ALL AuthProviderFactories to have a Catalog API** , which breaks or makes it much harder for users to implement their custom Authentication Mechanisms.

## Feature Suggestion
Remove or make implementing a CatalogAPI to the AuthProviderFactory optional.

## Possible Implementation

Given that I do not fully understand the motivation for this change, I would like to know why a CatalogAPI client is coupled to the Authentication Mechanism. 
Even if you feel such a coupling is justified, I would propose a change where in this coupling, is made optional `catalogApi?`, so that users can couple the CatalogAPI only if and when they need it.

**Comment by freben** [comment]
I JUST now made this one: https://github.com/backstage/backstage/issues/3870

TL;DR: The reasoning is that there needs to be a connection from the auth identity world to the backstage user/groups world. Somebody needs to make the connection. Currently the auth system is responsible for making the connection, even though it's - as is outlined in the RFC - not well fleshed out yet at all.

Give it a read if you have the time. I'm very happy to hear feedback on it in general, and to hear of alternatives.

**Comment by freben** [comment]
(and, you may enjoy the Alternatives section - which we talked a bit about and may actually lean towards - that renders the catalog connection unnecessary)

**Comment by GoWind** [comment]
Provided feedback on the RFC :). 

>Currently the auth system is responsible for making the connection, even though it's - as is outlined in the RFC - not well fleshed out yet at all.

I want to pushback a little (actually quite a bit) and say that this is better left to implementations rather than attempt to reason about this in the Core Backstage code/concepts. 

My motivation is even for our specific organization, we are still not sure how to model the relationship between GHE - Google and LDAP. We cannot decide on any connectiong mechanism without sync-ing with the rest of the teams handling this, which will be a long process. Also, I am pretty sure that any big enough organization will have the same bureaucratic and communication issues. 

These mappings might also be subject to change in the future and I would be happy to let the implementations decide on how to model this connection. 

A better way to solve this problem is to provide hooks to the frontends and the backends to select the information that they want to provide for fetching relevant entities. For example, with the Catalog UI plugin, in the default view, it would be better to provide an API , which a user can use to fetch the entities to be shown. An example of this would be , rather than filter and return all entities with `owner` set to `AuthProvider.Backstage.Identity.id` , if the UI providers an API `getComponentsByUserId(id: string)`, the UI customization can decide what entities to fetch.

**Comment by GoWind** [comment]
I would be happy to continue discussing this in the RFC, but as a solution for this specific issue, given that this coupling is still in the RFC stage and not decided yet, can we make the CatalogAPI an optional parameter in the interface ? That way, atleast until it is decided how to connect identity to Catalog, applications can decide for themselves how they want to do so.

**Comment by Rugvip** [comment]
One of the missing pieces is some sane way to enable or disable the identity part for each auth provider. We definitely don't want to force an identity mapping to the catalog for every single auth provider. The use-case for using multiple different auth providers as identity providers is for organizations that don't have all their users in the same SSO system or for other reasons want to use multiple ones. We still want providers like e.g. GitHub to only be used just for delegation of access to 3rd party resources.

The identity mapping we likely want in the auth backend in the end is `3rd party ID -> User Entity`, meaning we would be providing the catalog client to a "default mapping mechanism" for the providers rather than straight to the providers themselves, ultimately fixing what you're raising here. The user entity wouldn't necessarily have to exist within the catalog though, allowing organizations that can't fully map out their org structure in the catalog to still use some form of identity login, even though it would likely be a more limited form of login.

Custom mapping of ownership is tricky af though, since it touches on both frontend bits, backend logic like e.g. cost-insights, as well as authorization. Unless we want it to be purely aesthetic we'd likely want to model it as a separate service or endpoint. It's one of the reasons we're seeing the catalog as a good integration point, as it is a place were we can agree on the primitives for modelling things, and then creating a model that can be depended on directly rather than having to talk to a bunch of different surrounding services to get the full picture.

**Comment by GoWind** [comment]
> One of the missing pieces is some sane way to enable or disable the identity part for each auth provider. We definitely don't want to force an identity mapping to the catalog for every single auth provider. The use-case for using multiple different auth providers as identity providers is for organizations that don't have all their users in the same SSO system or for other reasons want to use multiple ones. We still want providers like e.g. GitHub to only be used just for delegation of access to 3rd party resources.

This is a non-problem, IMHO. If any organization does not have a unified identity system, then there is no way for me to recognize which user it is. 

For example, if my company uses Google SSO and LDAP , with half of my company using SSO and the rest using LDAP, then unless I have a mapping, that uniquely maps each ldap session or SSO session to a user, I cannot provide users uniform access to all resources.

If I allow different sources of authentication , then I will have to have an existing mapping from SSO, LDAP -> User.
In which case  this mapping
1. Might be the Catalog
2. Might be an already existing system
3. (or) present in the response from the authorization mechanism itself ( `idToken` for example). 

The approach currently assumes that this is in the Catalog or has to be in the Catalog. 


> The identity mapping we likely want in the auth backend in the end is 3rd party ID -> User Entity, meaning we would be providing the catalog client to a "default mapping mechanism" for the providers

I believe this mapping already exists in the frontend, where the Backstage Identity's `id` uses the information in the Id token. 

IIRC, this behaviour can be customized, by the `SignInProvider` (such as , using something other than the `idToken` to provide the Backstage User). 

Each time a user logs in via a certain authorization, the Backstage Identity changes. 

There cannot be multiple authorizations to a single user , as each authorization is Unique. 

Perhaps, we are modeling the wrong problem here. If I want a unified Backstage session across all of the different worlds (Private GHE, Public Github etc), it assumes that an organization has already mapped the different entites to a unified User. 

This is NOT universally true. 

My organization for example, has some notion of mapping Organization user -> Github.com user, but no such unified mapping exists, not does it seem there ever will. There are different rules that govern how work in OSS vs private work is done. 
In such a case, a unified catalog User is of no use to us. 

It is possible , that some organizations might use the Catalog to provide a unified mapping. But in such a case,
rather than an auth-backend providing the mapping, it is the Catalog that has to provide an API for mapping `Authorization` -> Backstage User and not the `auth-backend`. 

I think the dependencies are inverted here (which is my i want to ideally make this optional or remove the catalogApi).

**Comment by Rugvip** [comment]
> My organization for example, has some notion of mapping Organization user -> Github.com user, but no such unified mapping exists, not does it seem there ever will. There are different rules that govern how work in OSS vs private work is done.
> In such a case, a unified catalog User is of no use to us.

Does your "Organization user" have its own login provider and even organization data that you can ingest into the catalog?

**Comment by freben** [comment]
> IIRC, this behaviour can be customized, by the `SignInProvider` (such as , using something other than the `idToken` to provide the Backstage User).

Yeah the ID is "just a string" and can be sourced in any way, but the way that it is used after that, is lacking, and ties poorly to the ownership concepts.

> Each time a user logs in via a certain authorization, the Backstage Identity changes.
> 
> There cannot be multiple authorizations to a single user , as each authorization is Unique.

I was a bit unsure about these two - were they a hypothetical tied to the line above them, or a statement of the current state of things?

What the "labels" solution talked about, was similar to the existing one in the sense that yes, you return an ID on any form that makes sense to you, or rather, to your sign-in authority. Or multiple identities, even, perhaps! What Backstage then wants to be able to do, is to be able to find (optional) related data based on the provider key + that ID.

Am I understanding you correctly, that one way forward to you would be if the sign-in provider could be given the opportunity to return the object `{ google: 'john@example.com', ghe: 'john' }`? Where the second key was sourced out of some external system of yours where possible, but not always present?

> Perhaps, we are modeling the wrong problem here. If I want a unified Backstage session across all of the different worlds (Private GHE, Public Github etc), it assumes that an organization has already mapped the different entites to a unified User.

The second part of the alternative solution talks about not having any unified User, but rather to let each sign-in ID map to any number of Users, sourced from anywhere in any manner that fits your own org.

> It is possible , that some organizations might use the Catalog to provide a unified mapping. But in such a case,
> rather than an auth-backend providing the mapping, it is the Catalog that has to provide an API for mapping `Authorization` -> Backstage User and not the `auth-backend`.
> 
> I think the dependencies are inverted here (which is my i want to ideally make this optional or remove the catalogApi).

If the sign-in system (frontend or backend, wherever it happens) only emits a plain key, that may or may not map to a User, then maybe we get this loose coupling that we desire?

**Comment by GoWind** [comment]
> Does your "Organization user" have its own login provider and even organization data that you can ingest into the catalog?

Yes, we have our own login provider. 
We are currently discussing how to ingest organization data, because the setup is not straight forward and there is no way to synchronize cleanly with it. Which is why, we as of now, simply add `owner` info to the Entities, but do not store groups or users in the Catalog. 

@Rugvip

**Comment by GoWind** [comment]
> Yeah the ID is "just a string" and can be sourced in any way, but the way that it is used after that, is lacking, and ties poorly to the ownership concepts.

Yes, this is not a contract yet. The Backstage Identity and Authorization info is tied via `AuthResponse`, but there is no good contract between them. 

> I was a bit unsure about these two - were they a hypothetical tied to the line above them, or a statement of the current state of things?

I believe  this is the current state of things. `BackstageIdentity` is populated by the `populateIdentity` method, which is very specific to each of the Authentication providers. If you change  the  authorization provider, then this Identity changes based on the provider's `populateIdentity` method. It might be the same across the providers, but there is no guarantee and for therefore for all intents and purposes (unless you customize the populateIdentity method yourself), the same. 

>  What the "labels" solution talked about, was similar to the existing one in the sense that yes, you return an ID on any form that makes sense to you, or rather, to your sign-in authority. Or multiple identities, even, perhaps! What Backstage then wants to be able to do, is to be able to find (optional) related data based on the provider key + that ID.

I find this confusing (more precisely : `you return an ID on any form that makes sense to you, or rather, to your sign-in authority`)
Do you mean that I provide my sign-in authority with labels and my authority provides a `Backstage` Identity ? Or does my sign-in provider provide a label, that I then pass on to the catalog to return a `Backstage` User ? 

I do not for see an organization providing multiple ways to Sign-In to Backstage. Organizations may have wrapped their authentication systems with a Provider, such as Okta or Google, but I am effectively signing into my organization and viewing ONLY my organization's catalog. Therefore, there will be only one effective user at a time. 

If There is a an option to say, sign-in using my Github credentials, then I am longer using my organization's backstage application, but a backstage view into my Github, which I dont think is what organizations want. Even if you do, how relevant is an organization specific  plugin to your Github Identity in the Backstage Instance ? Should you even be seeing the plugin in your org's instance, when you are logged in as a Github user ? 

My reasoning is that the Identity returned by `populateIdentity` is the effective identity of my user and that there cannot be more than one at the same time. Currently even the session logic, IIRC is tied down to one specific user.

Google, for example has already solved this issue in the UI. You can use the same GMAIL to log into multiple accounts (you@gmail.com or you@your-company.com), but at any point , there is only ONE effective user. 

That said, a possible situation might be when 2 different orgs (org-a.com and org-b.com) share the same Backstage Instance and want users from both orgs to share the catalog view , but only modify based on the right org's ownership (only users from org A can view/modify entities belonging to org-A). 

For this, perhaps, having a `label` of sorts , might make sense . But I guess this is exactly where **`namespaces`** come into play, and this can be solved by adding a `namespace`.  The problem then become one of selecting the right namespace on the frontend, that  can then be passed onto the backend for filtering catalog entities.

**Comment by freben** [comment]
@GoWind I think we're actually converging in our thinking here. I agree with the part of just signing in using one method. Then the problem space is reduced to

- How may we tie the various strings in the little structs of data that comes out of our auth providers, to the `owner` fields of entities in the other end?
- How may we tie the signed-in persona to perhaps more than one such "space" of owner fields, that actually belong to other personas/groups in a foreign system?

We want to solve the first one, first and foremost. The second one is gravy but we're looking hungrily at that too.

For the first one, the idea was that "for google, it makes sense to use the email as the id". That's a good primary key.

OK, how do we go from that email, to what's in `owner` fields? We could put personal emails as owners but that's not where we want to be. We like group based ownership.

Ok, how do we go from the chosen primary key, to a list of strings that happen to be something group-like, and that occurs inside `owner` fields? As my latest comment in the RFC says, the sign-in system could provide a group list as part of its output - that's one possible way. Or we could have org data IN the catalog, that can be looked up based on that primary key. Or something else.

We believe that having org data in the catalog, and the associated functionality in the frontend that it unlocks, has massive value. It cannot be understated how much we appreciate to have that "full picture" internally of how things belong and interact and are owned and fit in a bigger context. And we attach tooling/plugins onto the users/groups themselves and make them searchable and navigable.

**Comment by Rugvip** [comment]
> > Does your "Organization user" have its own login provider and even organization data that you can ingest into the catalog?
> 
> Yes, we have our own login provider.
> We are currently discussing how to ingest organization data, because the setup is not straight forward and there is no way to synchronize cleanly with it. Which is why, we as of now, simply add `owner` info to the Entities, but do not store groups or users in the Catalog.

@GoWind wanna make sure it's clear that in your case we'd only then intend for that identity and org data to be ingested into the catalog. Other data such as GitHub would not be part of the org data model at all, unless you specifically want users to also be able to **sign in** to Backstage using GitHub, which would be a niche use-case.

In a typical setup you'd have that one source of org data with an accompanying auth and identity provider in the `auth-backend`. Other auth providers such as e.g. GitHub or Google auth are purely used for the Backstage users to be able to interact with those services from Backstage, but they are not tied to the org data or identity in any way.

That's what I was on about here:

> One of the missing pieces is some sane way to enable or disable the identity part for each auth provider. We definitely don't want to force an identity mapping to the catalog for every single auth provider. The use-case for using multiple different auth providers as identity providers is for organizations that don't have all their users in the same SSO system or for other reasons want to use multiple ones. We still want providers like e.g. GitHub to only be used just for delegation of access to 3rd party resources.

Regarding multiple sources of org data working together, we're also thinking the answer for that is namespaces.

**Comment by freben** [comment]
> For this, perhaps, having a `label` of sorts , might make sense . But I guess this is exactly where **`namespaces`** come into play, and this can be solved by adding a `namespace`. The problem then become one of selecting the right namespace on the frontend, that can then be passed onto the backend for filtering catalog entities.

Forgot to address this one.

We are looking at using namespaces as both scopes and possibly security boundaries. The way namespaces could play into what we suggest, is in terms of having different org trees in different namespaces to both avoid collisions and to scope them.

As for selecting namespaces in the frontend - that's not something we've been looking at so far. The plan has instead been to be able to set your scope in terms of a group or a user. Internally we have some of those kinds of breakdowns - "see how team X that you are in is doing" for possibly more than one X, if that's the case.

**Comment by GoWind** [comment]
@freben 
> How may we tie the various strings in the little structs of data that comes out of our auth providers, to the owner fields of entities in the other end?

`idTokens` are intended for this use-case. They contain the information that needed to authorize operations on behalf of the user. The backends task will be to obtain the `idToken`, verify it , unpack the information and use it to perform operations.It will by now have access to a user's group and other info that it can use for say, filter the catalog for entities, or for checking permissions, etc. 

The only pre-requisite is that org Information is ingested into the catalog for entities before hand. 

The `id` of the user is tied to the `idToken`. 

It is not easy to understand this, but implementations of ``AuthProviderRouteHandlers` be able to provide an id Token in the `AuthResponse` that is sent to the frontend. 

The frontend can use this information to carry out operations. 

There is no defined payload for an idToken. So, instead of providing APIs that can read specific fields, such as `user`, `groups`, etc, it would be simpler for the frontend to simply forward the idToken to the backend along with the requested operation.

Also this will solve the problem of authentication, as an idToken can also serve as `auth` Header. Currently, we had to setup extra protection via middleware to check for Auth Headers, as backstage's backend, by default, does not have any authentication mechanism.

**Comment by GoWind** [comment]
It is possible that `idTokens` are not sufficient. In which case, we possibly might have to provide a shim on the frontend to obtain a bag of additional info to enrich the frontend's session object. 

Note that this won't work well without the idToken, as a malicious request could "fake" this bag and we need to verify this. This is where access/idTokens can be used to request additional info on the frontend. 
Also, this bag of additional information has to be standardize across backend plugins so that they know where to look for. 

Maybe an optional method that the frontend calls post Sign In but Before creating session to obtain t his additional bag ? This can be configured via a config to be triggered along with an endpoint to make request to , to store this information ? 

Something like adding a `additional` Info to the `IdentityApi` , that returns an opaque object , which can be sent to the backend via a standardized header ?  

OTOH, thinking about it, I would avoid doing ^, as it just replicates what idToken does . Rather , if an idToken is insufficient, it should be up to the backend to make the additional call to obtain the info and carry out the operation.

**Comment by Rugvip** [comment]
@GoWind I could imagine making it possible to choose between either referencing an entity in the token, or embedding an entire entity. Obviously embedding an entire entity has some performance and synchronization issues, but would allow you to generate ad-hoc user data during login that doesn't necessarily exist in the catalog

**Comment by GoWind** [comment]
Yes, it is a trade-off. For context, this was a big multi-month discussion between multiple teams in a company I used to work for and I forsee the same problem occuring again :) (I had no idea what they concluded as i left before they did :)) . 
I think long-term having only Tokens might be the better solution. Backends might have to perhaps make additional requests, but 
 embedding entities in the frontend will only increase overtime as it is simpler ,because the problem is solved in one specific location , but in the long term might lead to more bloat. 
 
 I think this has to be perhaps an RFC ? I would like to know what the rest of the community thinks about this.

**Comment by Rugvip** [comment]
Yep, it's a discussion that'll have to continue into 2021 :grin:

**Comment by GoWind** [comment]
I think this issue can be closed ? I now see the need for the `CatalogApi`  and we have progressed on this topic with multiple RFCs and issues ..

**Comment by Rugvip** [comment]
@GoWind agreed, thanks!
